File: Algorithm/CPP/different_ways_to_add_parentheses.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;
class Solution {
public:
	vector<int> diffWaysToCompute(string input) {
		vector<int> vec;
		for (auto i = 0; i < input.size(); ++i)
		{
			if (ispunct(input[i]))
			{
				for (auto a : diffWaysToCompute(input.substr(0, i)))
				{
					for (auto b : diffWaysToCompute(input.substr(i + 1)))
					{
						if (input[i] == '+')
						{
							vec.push_back(a + b);
						}
						else if (input[i] == '-')
						{
							vec.push_back(a - b);
						}
						else
						{
							vec.push_back(a*b);
						}

					}

				}

			}
		}

		if (vec.size() == 0) vec.push_back(atoi(input.c_str()));
		return vec;

	}
};
int  main()
{
	Solution sol;
	string  input = "2*3-4*5";
	cout << input.c_str() << endl;
	auto vec= sol.diffWaysToCompute(input);
	for (auto k : vec)
		cout << k << endl;

	cin.get();
	return 0;
}

File: Algorithm/Python/different_ways_to_add_parentheses.py
import re
class Solution:
    def diffWaysToCompute(self, input):
        n = len(input)
        res = []
        for i in range(n):
            if input[i] in "+-*":
                for a in self.diffWaysToCompute(input[:i]):
                    for b in self.diffWaysToCompute(input[i+1:]) :
                        if input[i] == '+':
                            res.append(a+b)
                        elif input[i] == '-':
                            res.append(a-b)
                        else:
                            res.append(a*b)
        if not res:
            res.append(int(input))
        # get the operators
        return  res
        




if __name__ == '__main__':
    sol = Solution()
    input = "2*3-4*5"
    print sol.diffWaysToCompute(input)
